operator/internal/reconciliations/istio/installation.go

```diff
@@ -1,30 +1,99 @@
 package istio
 
 import (
+	"context"
+	"fmt"
 	operatorv1alpha1 "github.com/kyma-project/istio/operator/api/v1alpha1"
+	"github.com/kyma-project/istio/operator/internal/status"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 )
 
 type Installation struct {
-	Client         IstioClient
+	Client         LibraryClient
 	IstioVersion   string
 	IstioImageBase string
 }
 
 const (
 	LastAppliedConfiguration string = "operator.kyma-project.io/lastAppliedConfiguration"
+	installationFinalizer    string = "istios.operator.kyma-project.io/istio-installation"
 )
 
-// Reconcile runs Istio installation with merged Istio Operator manifest file when the trigger requires an installation.
-func (i *Installation) Reconcile(istioCR *operatorv1alpha1.Istio) error {
-	mergedIstioOperatorPath, err := merge(istioCR, i.Client.defaultIstioOperatorPath, i.Client.workingDir, TemplateData{IstioVersion: i.IstioVersion, IstioImageBase: i.IstioImageBase})
+// Reconcile runs Istio reconciliation to install, upgrade or uninstall Istio and returns the updated Istio CR.
+func (i *Installation) Reconcile(ctx context.Context, client client.Client, istioCR operatorv1alpha1.Istio, defaultIstioOperatorPath, workingDir string) (operatorv1alpha1.Istio, error) {
+
+	istioTag := fmt.Sprintf("%s-%s", i.IstioVersion, i.IstioImageBase)
+
+	// We need to evaluate what changed since last reconciliation, to make sure we run Istio reconciliation only if it's necessary
+	istioCRChanges, err := EvaluateIstioCRChanges(istioCR, istioTag)
 	if err != nil {
-		return err
+		ctrl.Log.Error(err, "Error evaluating IstioCR changes")
+		return istioCR, err
+	}
+
+	if !istioCRChanges.MustBeReconciled() {
```

Comment:
do we need this check at all?

Reply:
That's a good point, I think we can remove it.